### PR TITLE
[NIFI-13431] Initial nifi-jolt-transform-ui Nx project and initial ad…

### PR DESCRIPTION
**Pull request by scottyaslan**
…vanced UI layout/UX

Screenshot 2024-06-20 at 7 40 11 PM
Screenshot 2024-06-20 at 7 40 29 PM
Screenshot 2024-06-20 at 7 42 31 PM

### CODE REVIEW

**Comment by rfellows** [comment]
Will review...

File: nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import { joltTransformJsonUiFeatureKey } from '../state';
+
+describe('jolt-transform-json-ui', () => {
+    let component: JoltTransformJsonUi;
+    let fixture: ComponentFixture<JoltTransformJsonUi>;
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            declarations: [JoltTransformJsonUi],
+            imports: [RouterModule, RouterTestingModule],
```
Comment (rfellows) [inline nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.spec.ts:33]: Neither `RouterModule` nor `RouterTestingModule` seem to be needed to run this test.

File: nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                        }
+                    }
+                })
+            ]
+        });
+        fixture = TestBed.createComponent(JoltTransformJsonUi);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+        // @ts-ignore
+        window.js_beautify = () => '';
```
Comment (rfellows) [inline nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.spec.ts:48]: This doesn't appear to be needed at all to successfully run this test.

**Review by rfellows — CHANGES_REQUESTED** [review]
A few things inline. Also, there are license header check failures:
```
Files with unapproved licenses:

  src/main/frontend/libs/shared/jest.config.ts
  src/main/frontend/libs/shared/src/test-setup.ts
  src/main/frontend/libs/shared/src/assets/fonts/flowfont/flowfont.css
  src/main/frontend/libs/shared/src/assets/fonts/flowfont/flowfont.svg
  src/main/frontend/apps/nifi-jolt-transform-ui/proxy.config.mjs
```

Might need to update the files that get ignored if we can't add the proper header to them.

File: nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.ts [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+    validateJoltSpec
+} from '../state/jolt-transform-json-ui/jolt-transform-json-ui.actions';
+import { SavePropertiesRequest, ValidateJoltSpecRequest } from '../state/jolt-transform-json-ui';
+import { FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
+import { isDefinedAndNotNull } from 'libs/shared/src';
+
+const JS_BEAUTIFY_OPTIONS = {
+    indent_size: 1,
+    indent_char: '\t'
+};
```
[the remark this answers is not in the document]
Comment (scottyaslan) [reply nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.ts:50]: This is the settings used in the current Jolt Custom UI. However, I agree that if we update these settings as you propose we can then leverage the angular json pipe.

File: nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.html [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+                                            <i class="fa fa-align-left"></i>
+                                        </button>
+                                    </div>
+                                    <ngx-codemirror
+                                        formControlName="specification"
+                                        [options]="getJoltSpecOptions()"
+                                        (mousedown)="preventDrag($event)"></ngx-codemirror>
+                                </div>
+                                <div class="w-full overflow-ellipsis overflow-hidden whitespace-nowrap">
+                                    @if ((joltTransformJsonUiState$ | async)?.validatingJoltSpec) {
```
[the remark this answers is not in the document]
Comment (scottyaslan) [reply nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.html:97]: I have update to use signals here. The result is more readable as you requested.

File: nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.html [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+                                        <div class="input-editor flex-1">
+                                            <div class="flex flex-1 justify-end mb-4">
+                                                <button
+                                                    mat-icon-button
+                                                    color="primary"
+                                                    title="Format flow file content"
+                                                    (click)="formatInput()">
+                                                    <i class="fa fa-align-left"></i>
+                                                </button>
+                                            </div>
```
[the remark this answers is not in the document]
Comment (scottyaslan) [reply nifi-frontend/src/main/frontend/apps/nifi-jolt-transform-ui/src/app/pages/jolt-transform-json-ui/feature/jolt-transform-json-ui.component.html:253]: I see your point. However, in terms of information architecture I don't agree that the format mat-icon-buttons belong in the expansion panel header. It does appear that there is unused whitespace but we are actually following our established layout patterns. The button sits above the codemirror with `margin-bottom: 0.5rem;`. The button is the first thing in the DOM directly beneath the 'Jolt Specification' subheader. I think this is correct.

**Review by rfellows — APPROVED** [review]
This is a great first step in replacing the jolt transform custom ui. Thanks @scottyaslan.
I will merge once CI completes successfully.

**Pull request merged**